src/server/pfs/server/driver_file.go

```diff
@@ -263,13 +268,38 @@ func (d *driver) listFile(ctx context.Context, file *pfs.File, cb func(*pfs.File
 			})
 		}),
 	}
+	if paginationMarker != nil {
+		pathRange := &pfs.PathRange{}
+		if reverse {
+			pathRange.Upper = paginationMarker.Path
+		} else {
+			pathRange.Lower = paginationMarker.Path
+		}
+		opts = append(opts, WithPathRange(pathRange))
+	}
+	var fis []*pfs.FileInfo
 	s := NewSource(commitInfo, fs, opts...)
-	err = s.Iterate(ctx, func(fi *pfs.FileInfo, _ fileset.File) error {
+	if err = s.Iterate(ctx, func(fi *pfs.FileInfo, _ fileset.File) error {
 		if pathIsChild(name, pfsfile.CleanPath(fi.File.Path)) {
-			return cb(fi)
+			fis = append(fis, fi)
+			return nil
 		}
 		return nil
-	})
+	}); err != nil {
+		return errors.EnsureStack(err)
+	}
+	for i, fi := range fis {
```

Comment:
Doing this generally is going to be problematic for large lists. We need to change this algorithm to scale. I think we could do this by separating the forward from the reverse case, and have a separate callback that gets used for each. The forward case would just execute the callback above normally until the number has been emitted. The reverse case would collect a list up to a fixed size until the iteration completes and then return the number of files from that once the iteration completes. If the reverse list size gets too large, then we should return an error to prevent clients from OOM killing pachd. Also, it is worth noting the reverse performance can be O(n) towards the end of the lexicographical ordering of the file paths.